examples/psp22-extension/runtime/psp22-extension-example.rs

```diff
@@ -0,0 +1,289 @@
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{
+	log::{error, trace},
+	pallet_prelude::*,
+	traits::fungibles::{
+		approvals::{Inspect as AllowanceInspect, Mutate as AllowanceMutate},
+		Inspect, InspectMetadata, Transfer,
+	},
+};
+use pallet_assets::{self, WeightInfo};
+use pallet_contracts::chain_extension::{
+	ChainExtension, Environment, Ext, InitState, RetVal, SysConfig, UncheckedFrom,
+};
+use sp_runtime::DispatchError;
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+pub struct Psp22BalanceOfInput<AssetId, AccountId> {
+	pub asset_id: AssetId,
+	pub owner: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+pub struct Psp22AllowanceInput<AssetId, AccountId> {
+	pub asset_id: AssetId,
+	pub owner: AccountId,
+	pub spender: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+pub struct Psp22TransferInput<AssetId, AccountId, Balance> {
+	pub asset_id: AssetId,
+	pub to: AccountId,
+	pub value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+pub struct Psp22TransferFromInput<AssetId, AccountId, Balance> {
+	pub asset_id: AssetId,
+	pub from: AccountId,
+	pub to: AccountId,
+	pub value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+pub struct Psp22ApproveInput<AssetId, AccountId, Balance> {
+	pub asset_id: AssetId,
+	pub spender: AccountId,
+	pub value: Balance,
+}
+
+pub struct Psp22Extension;
+
+impl<T> ChainExtension<T> for Psp22Extension
+where
+	T: SysConfig + pallet_assets::Config + pallet_contracts::Config,
+	<T as SysConfig>::AccountId: UncheckedFrom<<T as SysConfig>::Hash> + AsRef<[u8]>,
+{
+	fn call<E: Ext>(func_id: u32, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+	where
+		E: Ext<T = T>,
+		<E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+	{
+		match func_id {
+			// Note: We use the PSP22 interface selectors as function IDs,
+			// there is no need but it makes sense from a convention perspective.
+
+			// PSP22 Metadata interfaces
+			0x3d261bd4 | 0x34205be5 | 0x7271b782 => {
+				let mut env = env.buf_in_buf_out();
+				let asset_id = env.read_as()?;
+				let result = match func_id {
+					// PSP22Metadata::token_name
+					0x3d261bd4 =>
+						<pallet_assets::Pallet<T> as InspectMetadata<T::AccountId>>::name(&asset_id)
+							.encode(),
+					// PSP22Metadata::token_symbol
+					0x34205be5 =>
+						<pallet_assets::Pallet<T> as InspectMetadata<T::AccountId>>::symbol(
+							&asset_id,
+						)
+						.encode(),
+					// PSP22Metadata::token_decimals
+					0x7271b782 =>
+						<pallet_assets::Pallet<T> as InspectMetadata<T::AccountId>>::decimals(
+							&asset_id,
+						)
+						.encode(),
+					_ => unreachable!(),
+				};
+				trace!(
+					target: "runtime",
+					"[ChainExtension] PSP22Metadata::{:?}",
+					func_id
+				);
+				env.write(&result, false, None).map_err(|_| {
```

Comment:
`None` is correct because the `result` is of fixed size. This means the weight for writing it is contained in the overall weight for the whole function.